### PR TITLE
Bypass the varint path when encoding InitMask

### PR DESCRIPTION
**Pull request by saethlin**
The data in a `InitMask` is stored as `u64` but it is a large bitmask (not numbers) so varint encoding doesn't make sense.

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs [hunk 1]
```diff
@@ -182,11 +184,33 @@ impl InitMask {
 /// The actual materialized blocks of the bitmask, when we can't keep the `InitMask` lazy.
 // Note: for performance reasons when interning, some of the fields can be partially
 // hashed. (see the `Hash` impl below for more details), so the impl is not derived.
-#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, HashStable)]
+#[derive(Clone, Debug, Eq, PartialEq, HashStable)]
 struct InitMaskMaterialized {
     blocks: Vec<Block>,
 }
 
+impl<E: TyEncoder> Encodable<E> for InitMaskMaterialized {
```
Comment (nnethercote) [inline compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:192]: Should `encode` and `decode` each have a brief comment explaining why they don't use `derive`?

Comment (nnethercote) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:197]: I'd put a comment on the decode impl that just refers to the encode impl :)

Comment (saethlin) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:197]: Done

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs [hunk 2]
```diff
@@ -182,11 +184,39 @@ impl InitMask {
...
 // Note: for performance reasons when interning, some of the fields can be partially
 // hashed. (see the `Hash` impl below for more details), so the impl is not derived.
-#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, HashStable)]
+#[derive(Clone, Debug, Eq, PartialEq, HashStable)]
 struct InitMaskMaterialized {
     blocks: Vec<Block>,
 }
 
+// `Block` is a `u64`, but it is a bitmask not a numeric value. If we were to just derive
+// Encodable and Decoable we would apply varint encoding to the bitmasks, which is slower
```
Comment (lqd) [inline compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:193]:
```suggestion
// Encodable and Decodable we would apply varint encoding to the bitmasks, which is slower
```

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs [hunk 3]
```diff
@@ -182,11 +184,39 @@ impl InitMask {
...
+// and also produces more output when the high bits of each `u64` are occupied.
+// Note: There is probably a remaining optimization for masks that do not use an entire
+// `Block`.
+impl<E: TyEncoder> Encodable<E> for InitMaskMaterialized {
+    fn encode(&self, encoder: &mut E) {
+        encoder.emit_usize(self.blocks.len());
+        for block in &self.blocks {
+            encoder.emit_raw_bytes(&block.to_le_bytes());
```
Comment (lqd) [inline compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:201]: Just wanted to note: I'm not familiar with the endianness rules for encoding/decoding, I had idly wondered before if we should prefer native endianness in incr data, but I think we use LE everywhere so 👍 .

**Review by lqd — APPROVED** [review]
Awesome.

~~I now wonder whether the const allocations' bytes themselves really need to be LEB128-encoded as well, and what would the time and size cost/gains be if we didn't.~~ (But probably they already aren't, they shouldn't and to be clear I haven't checked)

r=me with the typo suggestion squashed

**Comment by saethlin** [comment]
Largest regressions are in cranelift-codegen and keccak. Cachegrind diff for keccak looks like this (which should not surprise anyone):
```
1,374,605,175  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir             file:function
--------------------------------------------------------------------------------
1,374,460,232  ???:<rustc_data_structures::obligation_forest::ObligationForest<rustc_trait_selection::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection::traits::fulfill::FulfillProcessor>
```
This PR doesn't modify any of the related code. This looks like more optimization instability to me.

**Comment by nnethercote** [comment]
`keccak` and `cranelift-codegen` have just become bimodal again, with `keccak` giving an incredible +/- 8% swings. The other regressions (most notably `diesel`) are likely unrelated to that, and might be genuinely related to this PR.

**Comment by nnethercote** [comment]
On second thought, I just saw [this](https://github.com/rust-lang/rust/pull/110458#issuecomment-1512011390), so perhaps the `diesel` changes are just noise.

**Comment by saethlin** [comment]
The `diesel` diff looks like this:
```

199,979,359  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir             file:function
--------------------------------------------------------------------------------
  435,413,890  ???:<&rustc_middle::ty::list::List<rustc_middle::ty::subst::GenericArg> as rustc_type_ir::fold::TypeFoldable<rustc_middle::ty::context::TyCtxt>>::try_fold_with::<rustc_infer::infer::resolve::OpportunisticVarResolver>
 -263,175,486  ???:<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate> as rustc_type_ir::fold::TypeFoldable<rustc_middle::ty::context::TyCtxt>>::fold_with::<rustc_infer::infer::resolve::OpportunisticVarResolver>
 ```
 Which also seems pretty far away from what this PR is doing. So yeah, I agree.

**Comment by rylev** [comment]
Marking this as triaged since this was identified as most likely being noise.

@rustbot label: +perf-regression-triaged

**Pull request merged**